### PR TITLE
[REVIEW] Support for zero columned tables

### LINKED ISSUE #3047: [BUG] New cudf::table doesn't allow zero columned tables
**Describe the bug**
New cudf::table doesn't allow zero columned tables.

**Steps/Code to reproduce bug**
https://github.com/rapidsai/cudf/blob/4d0183133935d5ee0f06cd26658be1e77342988e/cpp/src/table/table.cpp#L36

**Expected behavior**
Should allow a table with zero columned table

**Environment overview (please complete the following information)**
 - Environment location: Bare-metal
 - Method of cuDF install: Source

### PR DESCRIPTION
**Pull request by rgsl888prabhu**
Support for zero columned tables

close #3047

### CODE REVIEW

File: cpp/src/table/table.cpp [hunk 1]
```diff
@@ -33,10 +33,13 @@ table::table(table const& other) : _num_rows{other.num_rows()} {
 // Move the contents of a vector `unique_ptr<column>`
 table::table(std::vector<std::unique_ptr<column>>&& columns)
     : _columns{std::move(columns)} {
-  CUDF_EXPECTS(columns.size() > 0, "Invalid number of columns");
-  _num_rows = columns[0]->size();
-  for (auto const& c : _columns) {
-    CUDF_EXPECTS(c->size() == num_rows(), "Column size mismatch.");
+  if(_columns.size() != 0) {
```
Comment (jrhemstad) [inline cpp/src/table/table.cpp]: Prefer to use public accessors when you can. 
```suggestion
  if(num_columns > 0) {
```

File: cpp/src/table/table.cpp [hunk 2]
```diff
@@ -33,10 +33,13 @@ table::table(table const& other) : _num_rows{other.num_rows()} {
...
+    _num_rows = _columns[0]->size();
+    for (auto const& c : _columns) {
```
Comment (jrhemstad) [inline cpp/src/table/table.cpp]: I realized that we should be doing extra safety checks and making sure that the contents of the unique_ptr are valid. 
```suggestion
    for (auto const& c : _columns) {
       CUDF_EXPECTS(c, "Unexpected null column");
```

File: cpp/src/table/table.cpp [hunk 3]
```diff
@@ -33,10 +33,13 @@ table::table(table const& other) : _num_rows{other.num_rows()} {
 // Move the contents of a vector `unique_ptr<column>`
 table::table(std::vector<std::unique_ptr<column>>&& columns)
     : _columns{std::move(columns)} {
-  CUDF_EXPECTS(columns.size() > 0, "Invalid number of columns");
-  _num_rows = columns[0]->size();
-  for (auto const& c : _columns) {
-    CUDF_EXPECTS(c->size() == num_rows(), "Column size mismatch.");
+  if(_columns.size() != 0) {
+    _num_rows = _columns[0]->size();
```
Comment (jrhemstad) [inline cpp/src/table/table.cpp]: Let's move this to after the loop.
```suggestion
```

File: cpp/src/table/table.cpp [hunk 4]
```diff
@@ -33,10 +33,13 @@ table::table(table const& other) : _num_rows{other.num_rows()} {
...
+    for (auto const& c : _columns) {
+      CUDF_EXPECTS(c->size() == num_rows(), "Column size mismatch.");
+    }
```
Comment (jrhemstad) [inline cpp/src/table/table.cpp]:
```suggestion
    }
     _num_rows = _columns[0]->size();

```

File: cpp/src/table/table.cpp [hunk 5]
```diff
@@ -33,10 +33,15 @@ table::table(table const& other) : _num_rows{other.num_rows()} {
...
     : _columns{std::move(columns)} {
-  CUDF_EXPECTS(columns.size() > 0, "Invalid number of columns");
-  _num_rows = columns[0]->size();
-  for (auto const& c : _columns) {
-    CUDF_EXPECTS(c->size() == num_rows(), "Column size mismatch.");
+  if(num_columns() > 0) {
+    _num_rows = _columns[0]->size();
+    for (auto const& c : _columns) {
+      CUDF_EXPECTS(c, "Unexpected null column");
+      CUDF_EXPECTS(c->size() == num_rows(), "Column size mismatch.");
```
Comment (jrhemstad) [inline cpp/src/table/table.cpp]: I realized we'll need to change this to make my other suggestion work. 
```suggestion
      CUDF_EXPECTS(c->size() == _columns.first(), "Column size mismatch.");
```

Comment (rgsl888prabhu) [reply cpp/src/table/table.cpp]: Ah, I missed it as well.

File: cpp/src/table/table.cpp [hunk 6]
```diff
@@ -33,10 +33,15 @@ table::table(table const& other) : _num_rows{other.num_rows()} {
 // Move the contents of a vector `unique_ptr<column>`
 table::table(std::vector<std::unique_ptr<column>>&& columns)
     : _columns{std::move(columns)} {
-  CUDF_EXPECTS(columns.size() > 0, "Invalid number of columns");
-  _num_rows = columns[0]->size();
-  for (auto const& c : _columns) {
-    CUDF_EXPECTS(c->size() == num_rows(), "Column size mismatch.");
+  if(num_columns() > 0) {
+    _num_rows = _columns[0]->size();
```
Comment (jrhemstad) [inline cpp/src/table/table.cpp]:
```suggestion
```
This should be removed.

**Comment by jrhemstad** [comment]
rerun tests

**Comment by rgsl888prabhu** [comment]
rerun tests

**Comment by harrism** [comment]
I wouldn't bother rerunning tests until we resolve all the RMM issues.

**Pull request merged**